ndcube/ndcube.py

```diff
@@ -409,21 +419,46 @@ def extra_coords(self):
                     "value": self._extra_coords_wcs_axis[key]["value"]}
         return result
 
-    def crop_by_coords(self, min_coord_values, interval_widths):
+    def crop_by_coords(self, lower_corner, interval_widths=None, upper_corner=None):
         # The docstring is defined in NDDataBase
 
         n_dim = len(self.dimensions)
-        if (len(min_coord_values) != len(interval_widths)) or len(min_coord_values) != n_dim:
-            raise ValueError("min_coord_values and interval_widths must have "
-                             "same number of elements as number of data dimensions.")
-        # Convert coords of lower left corner to pixel units.
-        lower_pixels = self.world_to_pixel(*min_coord_values)
-        upper_pixels = self.world_to_pixel(*[min_coord_values[i] + interval_widths[i]
-                                             for i in range(n_dim)])
-        # Round pixel values to nearest integer.
-        lower_pixels = [int(np.rint(l.value)) for l in lower_pixels]
-        upper_pixels = [int(np.rint(u.value)) for u in upper_pixels]
-        item = tuple([slice(lower_pixels[i], upper_pixels[i]) for i in range(n_dim)])
+        # Raising a value error if the arguments have not the same dimensions.
+        if upper_corner:
+            if (len(lower_corner) != len(upper_corner)) or (len(lower_corner) != n_dim):
+                raise ValueError("lower_corner and upper_corner must have "
+                                 "same number of elements as number of data "
+                                 "dimensions.")
+        # Raising a value error if the arguments have not the same dimensions.
+        # Calculation of upper_corner with the inputing interval_widths
+        # This part of the code will be removed in version 2.0
+        if interval_widths:
+            warnings.warn("interval_widths will be removed from the API in "
+                          "version 2.0, please use upper_corner argument.")
+            if (len(lower_corner) != len(interval_widths)) or (len(lower_corner) != n_dim):
+                raise ValueError("lower_corner and interval_widths must have "
+                                 "same number of elements as number of data "
+                                 "dimensions.")
+            upper_corner = [lower_corner[i] + interval_widths[i] for i in range(n_dim)]
```

Comment:
I think it's better if you reverse the order of these ```if``` statements.  That way you can issue the warning and construct ```upper_corner``` from ```interval_widths``` without checking the number of elements is right.  Then you can just check the length of ```upper_corner``` without requiring an ```if``` statement since if it wasn't defined by the user, you will have defined in the ```if interval_widths:``` if statement.

Then when it comes time to stop supporting the ```interval_widths``` API, we just have to delete the ```if interval_widths:``` statement.

Reply:
If i delete the elements number checking between `lower_corner`, `interval_widths` and `n_dim`, an error will come by creating the `upper_corner` values because one of these tuples will be out of range. Do you want I remove this error raising ?

Reply:
Ah, you're absolutely right!  Yes, leave in the length checking in both statements, but check ```interval_widths``` first.  Then you can also add in a check that ```interval_widths``` and ```upper_corner``` were not both set:
```python
# Calculation of upper_corner with the inputing interval_widths
# This part of the code will be removed in version 2.0
if interval_widths:
    warnings.warn("interval_widths will be removed from the API in "
                  "version 2.0, please use upper_corner argument.")
    # If both intervals_widths and upper_corner set, raise error.
    if upper_corner:
        raise ValueError("Only one of interval_widths or upper_corner can be set.  Recommend using upper_corner as interval_widths is deprecated.")
    if (len(lower_corner) != len(interval_widths)) or (len(lower_corner) != n_dim):
        raise ValueError("lower_corner and interval_widths must have "
                         "same number of elements as number of data "
                         "dimensions.")
    upper_corner = [lower_corner[i] + interval_widths[i] for i in range(n_dim)]
# Raising a value error if the arguments have not the same dimensions.
if (len(lower_corner) != len(upper_corner)) or (len(lower_corner) != n_dim):
                raise ValueError("lower_corner and upper_corner must have "
                                 "same number of elements as number of data "
                                 "dimensions.")

Reply:
Ok, I will input this line in my code, I already exchanged the if statements positions.